text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,100 @@
+```
+Modifications will be made to mark the method above as deprecated and another method created to remove the wildcard which will allow the subscription pattern to look like below
+>folio\.Default\.DI_COMPLETED
+
+Safeguards will be added to ensure that a message has a tenant id as one of its headers. It is imperative that every message has a owner.
+
+### Producing
+In folio-kafka-wrapper, a topic name is generated by this interface
+```
+public static String formatTopicName(String env, String nameSpace, String tenant, String eventType)
+```
+This will generate a sample of a subscription pattern like so:
+>folio\.Default\.tenant00001\.DI_COMPLETED
+
+Modifications will be made to mark this method as deprecated and another method is created to remove the tenant id reference. This will generate a sample of a subscription pattern like so:
+>folio\.Default\.DI_COMPLETED
+
+Deprecated methods will allow an easier transition as processes roll off tenant specific topics to single topics.
+Safeguards will be added to ensure that a message has a tenant id as one of its headers. It is imperative that every message has a owner.
+## Migration
+- Changes will need to be made in folio-kafka-wrapper to stop creating topics for each tenant.
+- FOLIO modules will need to reference the latest version of folio-kafka-wrapper.
+- During a flower release, there should be no usage occurring in a FOLIO cluster. Existing topics will be deleted. Care should be taken to ensure all messages in the topic is consumed before topic deletion.
+- FOLIO modules are instantiated, new topics are created.
+- Migration can happen on a topic basis i.e a set of topics are migrated at a time instead of all topics within a module. What most important is identifying the consumers and producers of a topic to make necessary updates to allow proper topic identification, production & consumption.
+
+
+## Risks and Drawbacks
+- FOLIO's multi-versioning scheme involving tenants using different versions of modules will cause functional issues. Multiple versions of a module will be consuming from the same topic and there are no guarantees about which module version will consume a message sourced from an incompatible version.
+- Insights like "how many messages have yet to be processed for a tenant" will be harder to derive.
+- Per tenant topic settings will not possible.
+- [Temporary Kafka Security Solution](https://wiki.folio.org/x/YYVFAw) will no longer be possible. This solution is meant to be temporary as its name implies, a stop gap to a permanent solution. Creating topics for each tenant has been implemented and lightly enforced via the [folio-kafka-wrapper](https://github.com/folio-org/folio-kafka-wrapper) library. FOLIO is not able to enforce tenant security and isolation within its boundaries: the other two parts are not defined within FOLIO code, it is up to the hosting provider to implement. 
```

Comment:
> I see Kafka as a transport channel instead of a repository for data. 

I can understand this perspective. And that might be true of how FOLIO uses Kafka. I don't think that's necessarily true of Kafka in general.

> Multi-tenancy in FOLIO does not mean separate module instances or networking infrastructure for each tenant. It means multiple tenants should be able to co-exists and function properly without data contamination. 

Ok.

> The current structure from a strict data isolation perspective does not hold true because it is like having hotel rooms without doors or locks. 

I'm not sure I understand the analogy.

> The current structure organizes data but does not isolate. More work will need to be on FOLIO modules to get true data isolation.

Ok, I can accept that the current implementation might not reach our aspirations. 

How is that relevant to the feedback I gave?

My interpretation of what you are saying is that because FOLIO already doesn't strictly isolate tenant data to the highest degree or at all levels, then it's ok to actively choose a design which does not allow for the isolation at all.

How wrong is my interpretation?

Reply:
>I'm not sure I understand the analogy.

Ha. Maybe the missing piece of information is that there is an entrance to each room but it has no door. Patrons are able to move between rooms they like.

>How is that relevant to the feedback I gave?

My comment is meant to illustrate that similar level of standard is maintained.

>My interpretation of what you are saying is that because FOLIO already doesn't strictly isolate tenant data to the highest degree or at all levels, then it's ok to actively choose a design which does not allow for the isolation at all.

The current design can be extended for true isolation but not with the approach listed by the "temporary solution". We have chosen to keep Kafka security out of this design and an approach I have in mind can add true isolation. I am not sure if the community has the will to develop this further since other levels are not achieving the desired level of isolation and the "temporary solution" was not fully implemented. Unlike other technical issues with FOLIO, this issue is costing hosting providers every billing cycle hence why I believe the bleeding needs to be stopped. With this completion of implementation of this RFC, tackling kafka security would be the next step. I am open to chat with you about my idea kafka security offline!

Reply:
> Ha. Maybe the missing piece of information is that there is an entrance to each room but it has no door. Patrons are able to move between rooms they like.

Unfortunately that does not help me at all.

I think you are arguing that you do not consider schema separation in the database to be sufficient isolation. I think this is another example that I might not understand what you mean by isolation.

> My comment is meant to illustrate that similar level of standard is maintained.

The current standard in the database is to actively (logically) separate the data for each tenant, by using separate schema. 

To me, the closest analogy to that is separate topics. Using the same topic and a header would be similar to using a single schema and each table having a tenant discriminator column.

> The current design can be extended for true isolation

What do you mean by true isolation? I keep getting the sense that I'm getting caught up on what you mean by that word, sometimes I think you have a stricter standard than I do, other times weaker.

I don't understand how it could be possible to achieve stronger isolation once the data for multiple tenants is actively put into a single topic. Maybe I'm missing something obvious.

> We have chosen to keep Kafka security out of this design

What do you mean by security in this context? Access control I can understand because that has been deferred multiple times.


> Unlike other technical issues with FOLIO, this issue is costing hosting providers every billing cycle hence why I believe the bleeding needs to be stopped. 

I understand why this is important from that perspective. As hosting is entirely separated from development in FOLIO, I likely have a different perspective.

And I think that if that is the primary concern, I think there are ways to improve that situation significantly with less work than this proposal.

Reply:
I agree that my standard for isolation seems like it is shifting. I think it is because I am maintaining the status quo in some aspects and questioning the status quo in others. As the owner of this RFC, I have had to confront the promise of FOLIO's architecture when it was first designed and its current reality.

The current level of isolation at the database level is sufficient enough to enhance querying of data and administrative abilities. This could be substituted with table partitioning to derive these similar benefits. Where it lacks is preventing unauthorized access to data; the same identity is used to access all tenants.

The current level of isolation in kafka is sufficient and even great if the temporary kafka security solution is implemented.  But because Kafka's current usage is as a transport mechanism, I treat it a little different. No one would push for separate networking in the name of isolation for FOLIO especially since, in general, it is consumed by the same module instance. Pushing for module instances for each tenant is very cost prohibitive business-wise and a non-starter from an operations perspective when you would like 10s or 100s of tenants in one cluster with more than 150 module types. Secure procurement & management of numerous identities for postgres/kafka is not trivial. So let's reimagine isolation not like it is currently implemented in the database. Multiple tenant packets flowing through the same network but with encryption to hide data for others. Similarly multiple tenant messages in a topic with encryption to hide data from others. There is no expectation to query data from the middle of a network path. Similarly there is no expectation to query from the middle of topic. First In, First Out.